### PR TITLE
ARROW-623: Fix segfault in __repr__ of empty field

### PR DESCRIPTION
**Pull request by tebeka**
Small fix for not segfaulting when printing an empty field.

### CODE REVIEW

File: python/pyarrow/tests/test_schema.py [hunk 1]
```diff
@@ -87,3 +87,10 @@ def test_schema_equals(self):
         del fields[-1]
         sch3 = A.schema(fields)
         assert not sch1.equals(sch3)
+
+
+class TestField(unittest.TestCase):
+    def test_empty_field(self):
+        f = arrow.Field()
+        # This used to segfault on empty field
+        repr(f)
```
Comment (wesm) [inline python/pyarrow/tests/test_schema.py]: Using the `Field` constructor direct is not supported -- can you have this instead raise an exception?

**Comment by wesm** [comment]
Can you open a JIRA and prefix the PR title with `ARROW-XXX:`?

**Comment by tebeka** [comment]
Opened https://issues.apache.org/jira/browse/ARROW-623

File: python/pyarrow/tests/test_schema.py (hunk 1 above)
Comment (tebeka) [inline python/pyarrow/tests/test_schema.py]: `Field` is constructed in two sages (schema.pyx:131):
{code}
        cdef Field result = Field()
        result.init(self.schema.field(i))
{code}

And there's also `Field.from_py` which also creates an empty field.

How can I tell between these cases to ones where this is not OK. Maybe remove the `Field` import from pyarrow `__init_.py`?

Comment (wesm) [inline python/pyarrow/tests/test_schema.py]: This is an inherent limitation of Cython extension types, that their constructors are not generally safe to use directly. So when I said "is not supported" I meant, "is not supported as a public API". The other things you cited are "private API" usages. If you look at the other extension types that wrap other Arrow types (DataType, Array, Schema, etc.), you can see they all follow the empty-ctor followed by init-from-shared_ptr

Comment (tebeka) [inline python/pyarrow/tests/test_schema.py]: You said
> can you have this instead raise an exception?

By "this" your mean the repr? (I though you meant the creation)

**Comment by wesm** [comment]
Bottom line: the statement `pyarrow.Field()` is not supported for users -- it creates a malformed object. The class docstring should bear a warning like https://github.com/apache/arrow/blob/master/python/pyarrow/table.pyx#L323

File: python/pyarrow/tests/test_schema.py (hunk 1 above)
Comment (wesm) [inline python/pyarrow/tests/test_schema.py]: The repr

Comment (wesm) [inline python/pyarrow/tests/test_schema.py]: In some other classes we have a convenience function for this: https://github.com/apache/arrow/blob/master/python/pyarrow/table.pyx#L63

**Comment by wesm** [comment]
Looks like the build stalled for some other reason. Merging, thanks!

**Pull request closed without merging**